Feature/task11

HubertKarw opened the pull request.

HubertKarw committed:
java/product-project/task11/thread impl

HubertKarw committed:
java/product-project/task11/removed unnecessary code

HubertKarw committed:
Merge branch 'master' into feature/task11

kkpolak commented on src/com/company/OrderProcessor.java:
```diff
@@ -53,6 +57,17 @@ public static void createFile() {
+        orderToProcess.markOrderAsProcessed();
+        String receipt = createReceipt();
+        System.out.println(receipt);
+        writeToFile(receipt);
```
Kiedy wiele wątków jednocześnie wywołuje writeToFile(receipt), może dojść do „wymieszania” danych w pliku, bo kolejne wątki przeplotą się (interleave anomalia) na poziomie zapisu.
Co bym zalezal:

- Najprostsza opcja: metody createFile() i writeToFile() mogłyby być synchronized, żeby w danym czasie tylko jeden wątek pisał do pliku.
- Alternatywnie możesz użyć mechanizmu blokad (np. ReentrantLock) albo kolejki, gdzie parawony (treści do zapisania) będą „wrzucane”, a osobny wątek będzie je pobierał i zapisywał do pliku.

kkpolak commented on src/com/company/OrderProcessor.java:
```diff
@@ -7,10 +7,18 @@
 import java.time.format.DateTimeFormatter;
```
ogolnie rozwiazanie w porzadku, ale jesli chcemy wejsc na wyzszy poziom to w kodzie produkcyjnym na projekcie takie cos raczej byloby rozwiazane przy pomocy tych nowszych bardziej zaawansowanych narzedzi, jesli masz czas i ochote to mozesz sprobowac pobawic sie z ExecutorService:

- Utworzenie puli wątków (np. ExecutorService executor = Executors.newFixedThreadPool(4);)
- Dodawanie nowych zamówień do kolejki zadań (executor.submit(() -> new OrderProcessor(order).run()))
- W ten sposób łatwiej skalować liczbę wątków. Gdy zamówień będzie bardzo dużo, nie zalewasz systemu setkami wątków, tylko pracujesz w stabilnej puli, która jest wielokrotnie wykorzystywana.

HubertKarw committed:
java/product-project/task11/synchronized file methods and executor service
added synchronized keywoerd to createFile and writeToFile methods
created executor service in OrderProcessor

HubertKarw committed:
java/product-project/task11/test case
added multipleOrdersAtTheSameTime method for testing the executorService in order processor

HubertKarw committed:
java/product-project/task11/executor changes
